agent/job_runner.go

```diff
@@ -517,18 +526,23 @@ func (r *JobRunner) startJob(startedAt time.Time) error {
 
 // Finishes the job in the Buildkite Agent API. This call will keep on retrying
 // forever until it finally gets a successfull response from the API.
-func (r *JobRunner) finishJob(finishedAt time.Time, exitStatus string, signal string, failedChunkCount int) error {
+func (r *JobRunner) finishJob(finishedAt time.Time, exitStatus string, signal string, signalReason string, failedChunkCount int) error {
 	r.job.FinishedAt = finishedAt.UTC().Format(time.RFC3339Nano)
 	r.job.ExitStatus = exitStatus
 	r.job.Signal = signal
 	r.job.ChunksFailedCount = failedChunkCount
 
-	// If the agent has been stopped, send a signal reason of `agent_stop` to distinguish between
-	// user-generated cancels and those due to the agent getting an operating system signal. If
-	// the job was signalled because it was cancelled then the reason is `cancel`.
-	if r.stopped {
+	if signalReason != "" {
+		r.job.SignalReason = signalReason
+	} else if r.stopped {
+		// If the agent has been stopped, send a signal reason of
+		// `agent_stop` to distinguish between user-generated cancels
+		// and those due to the agent getting an operating system
+		// signal.
 		r.job.SignalReason = `agent_stop`
 	} else if r.cancelled {
```

Comment:
It's not really a problem with this PR, but can this branch ever be true? I can't see where `r.cancelled` is set to True

Reply:
No idea :-)

Reply:
Looks like it should probably be set to true somewhere in `func (r *JobRunner) Cancel() error`

Reply:
yup, PR incoming

Reply:
looks like we stopped setting `r.cancelled` in 6275d45e, as part of #882

Reply:
It looks like `r.cancelled = true` was removed in 6275d45e4eda1111b326c71668ab76120fbd6ff3, I think perhaps by mistake

Reply:
I think setting it in `func (r *JobRunner) Cancel() error` will work, but then when `func (r *JobRunner) CancelAndStop()` is called as part of graceful agent shutdown we'll end up with both `stopped` AND `cancelled` set to `True` :thinking:

Reply:
I pushed up a quick commit which I think will fix.

Reply:
I'd be tempted to move the signalReason calculation up into `Run` instead of having it within finishJob, tbqh.

Reply:
Agreed

Reply:
Oh sorry @yob, I didn't see "yup, PR incoming" until I re-read this from a notification. I think GitHub wasn't rendering it in page until I did a proper refresh.

Feel free to pop that commit off here and do it separately.